README.md
# SlideDrawer

An easy way to use drawer in Flutter with cool sliding animation.

![](example.gif)

## Getting Started

Add dependency in your flutter project.

```
$ flutter pub add slide_drawer
```

or

```yaml
dependencies:
  slide_drawer: ^1.0.3
```

Then run `flutter pub get`.

## Example

You can see the `example` folder to learn many different ways of using SlideDrawer in your app.

## Usage

As usual, we need to import the package.

```dart
import 'package:slide_drawer/slide_drawer.dart';
```

Then wrap the app home page with SlideDrawer. 

To use the basic SlideDrawer, you only need to define items (List of MenuItem) to generate menu in your drawer. The drawer will use your default theme color as drawer background, and theme brightness as drawer brightness.

```dart
class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: title,
      // SlideDrawer will use style from this theme
      theme: ThemeData(
        brightness: Brightness.dark,
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      // wrap your HomePage with SlideDrawer
      home: SlideDrawer(
        items: [
          MenuItem('Home', onTap: (){}),
          MenuItem('Project', onTap: (){}),
          MenuItem('Favourite', onTap: (){}),
          MenuItem('Profile', onTap: (){}),
          MenuItem('Setting', onTap: (){}),
        ],
        child: HomePage(),
      ),
    );
  }
}
```

Then set onPressed in your AppBar menu button to toggle SlideDrawer.

```dart
appBar: AppBar(
  title: Text(title),
  leading: IconButton(
    icon: Icon(Icons.menu),
    // call toggle from SlideDrawer to alternate between open and close
    // when pressed menu button
    onPressed: () => SlideDrawer.of(context).toggle(),
  ),
),
```

### Icons

We can also use icon for menu items.

```dart
home: SlideDrawer(
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  child: HomePage(),
),
```

### Custom Background

SlideDrawer also come with parameter backgroundColor and backgroundGradient. 

If we specify backgroundGradient, the drawer will use it as background. If not, it will use backgroundColor. If we also not specify backgroundColor, it will use the default theme color as background.

It's good practice to specify brightness (dark or light) if we use backgroundGradient or backgroundColor. 
If we set dark color in your backgroundGradient or backgroundColor, use Brightness.dark.
If we set light color, use Brightness.light.

```dart
// SlideDrawer with custom gradient as background.
home: SlideDrawer(
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  brightness: Brightness.dark,
  backgroundGradient: LinearGradient(
    begin: Alignment.topCenter,
    end: Alignment.bottomCenter,
    stops: [0.0, 1.0], 
    colors: [
      Color(0xFF000046),
      Color(0xFF1CB5E0),
    ],
  ),
  child: HomePage(),
),
```

```dart
// SlideDrawer with custom color as background.
home: SlideDrawer(
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  brightness: Brightness.dark,
  backgroundColor: Colors.blue[900],
  child: HomePage(),
),
```

### Custom Head Drawer

It's advisable to set alignment to SlideDrawerAlignment.start when you use headDrawer.

```dart
home: SlideDrawer(
  alignment: SlideDrawerAlignment.start,
  headDrawer: Image.asset('boys.png'),
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  brightness: Brightness.dark,
  backgroundColor: Colors.blue[900],
  child: HomePage(),
),
```

### Custom Content Drawer

In the previous example, we use items and let the SlideDrawer generate menu in the content drawer.
But we can also set any custom widget as the content drawer.

```dart
home: SlideDrawer(
  brightness: Brightness.dark,
  contentDrawer: Container(
    padding: EdgeInsets.symmetric(horizontal: 5),
    child: Column(
      children: [
        ListTile(title: 'Home', icon: Icons.home, onTap: (){}),
        ListTile(title: 'Project', icon:Icons.rss_feed, onTap: (){}),
        ListTile(title: 'Favourite', icon: Icons.favorite_border, onTap: (){}),
        ListTile(title: 'Profile', icon: Icons.person_outline, onTap: (){}),
        ListTile(title: 'Setting', icon: Icons.settings, onTap: (){}),
      ],
    ),
  ),
  child: HomePage(),
),
```

### Custom Full Drawer

If you have your own creative idea for the full widget inside the drawer, don't worry. We can use it in  SlideDrawer by passing it to drawer. In this way, SlideDrawer will use style of custom drawer and ignore any value in backgroundColor, gradientColor, and brightness.

```dart
home: SlideDrawer(
  drawer: Container(
    color: Colors.teal,
    padding: EdgeInsets.symmetric(vertical: 36, horizontal: 15),
    child: Theme(
      data: ThemeData(brightness: Brightness.dark),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.end,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          ListTile(title: 'Home', icon: Icons.home, onTap: (){}),
          ListTile(title: 'Project', icon:Icons.rss_feed, onTap: (){}),
          ListTile(title: 'Favourite', icon: Icons.favorite_border, onTap: (){}),
          ListTile(title: 'Profile', icon: Icons.person_outline, onTap: (){}),
          ListTile(title: 'Setting', icon: Icons.settings, onTap: (){}),
        ],
      ),
    ),
  ),
  child: HomePage(),
),
```

### Others

The default SlideDrawer animate home page by move it to the right (offsetFromRight default to 60), scale it, and rotate it at y axis with some small angle (pi / 24 radian or 180 / 24 degree = 7.5 degree). 

If don't want to rotate, we can disable the rotation by setting isRotate to false.

```dart
home: SlideDrawer(
  isRotate: false,
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  child: HomePage(),
),
```

Or use a custom rotateAngle if we want to rotate it at a spesific angle.

```dart
home: SlideDrawer(
  isRotate: true,
  rotateAngle: pi / 36,
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  child: HomePage(),
),
```

Or change offsetFromRight to make it move a little closer or farther.

```dart
home: SlideDrawer(
  offsetFromRight: 80.0,
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  child: HomePage(),
),
```

We can also use custom curve and duration. The default values are linear and 300ms. We can specify curveReverse and durationReverse to change curve and duration of animation in the reverse direction.

```dart
home: SlideDrawer(
  curve: Curves.easeInOut,
  duration: Duration(milliseconds: 200),
  items: [
    MenuItem('Home', icon: Icons.home, onTap: (){}),
    MenuItem('Project', icon:Icons.rss_feed, onTap: (){}),
    MenuItem('Favourite', icon: Icons.favorite_border, onTap: (){}),
    MenuItem('Profile', icon: Icons.person_outline, onTap: (){}),
    MenuItem('Setting', icon: Icons.settings, onTap: (){}),
  ],
  child: HomePage(),
),
```

There it is. Have fun!!!

CHANGELOG.md
## [1.0.3] - 22 June 2021

* auto close when tap menu item

## [1.0.2] - 19 May 2021

* Migrate to null safety.

## [1.0.1] - 18 Sept 2020

* Upgrade curved_animation_controller.

## [1.0.0] - 11 Sept 2020

* A working slide drawer.


lib/src/drawer.dart
import 'dart:math';

import 'package:curved_animation_controller/curved_animation_controller.dart';
import 'package:flutter/material.dart';
import 'package:slide_drawer/src/alignment.dart';
import 'package:slide_drawer/src/container.dart';
import 'package:slide_drawer/src/item.dart';

class SlideDrawer extends StatefulWidget {
  /// The gradient to use for the drawer background.
  ///
  /// If this is null, then [background] is used.
  final Gradient? backgroundGradient;

  /// The color to use for the drawer background. Typically this should be set
  /// along with [brightness].
  ///
  /// If this is null, then [ThemeData.primaryColor] is used.
  final Color? backgroundColor;

  /// The brightness of the drawer. Typically this is set along
  /// with [backgroundColor], [backgroundGradient].
  ///
  /// If this is null, then [ThemeData.primaryColorBrightness] is used.
  final Brightness? brightness;

  /// This is where you should attach your main application widget
  final Widget child;

  /// Custom drawer to be used when you don't want to use
  /// the default [SlideDrawerContainer] generated from [items] or [contentDrawer]
  final Widget? drawer;

  /// Head drawer to be rendered before [contentDrawer]
  /// or the default generated content drawer from [items]
  final Widget? headDrawer;

  /// Content drawer to be used if you don't want to use
  /// the default content drawer generated from [items]
  final Widget? contentDrawer;

  /// List of [MenuItem] to be used to generate the default content drawer
  final List<MenuItem> items;

  /// Duration of the drawer sliding animation
  ///
  /// Default is [Duration(milliseconds: 300)]
  final Duration duration;

  /// Curve to be used for the drawer sliding animation
  final Curve curve;

  /// Duration of the drawer sliding animation in the reverse direction
  final Duration? reverseDuration;

  /// Curve to be used for the drawer sliding animation in the reverse direction
  final Curve? reverseCurve;

  /// Vertical alignment of content inside drawer
  /// it can [start] from the top, or [center]
  final SlideDrawerAlignment? alignment;

  /// Offset from right to calculate the end point of sliding animation
  ///
  /// Default is 60.0
  final double offsetFromRight;

  /// Whether you want to rotate the [child] in the sliding animation
  ///
  /// Default is [true]
  final bool isRotate;

  /// Rotation angle of the [child] (radian) when
  /// the [SlideDrawer] sliding to the right
  ///
  /// Default is [pi / 2]
  final double rotateAngle;

  final Function? onWillPop;

  const SlideDrawer({
    Key? key,
    this.items = const [],
    this.drawer,
    this.headDrawer,
    this.contentDrawer,
    required this.child,
    this.backgroundGradient,
    this.backgroundColor,
    this.brightness,
    this.duration = const Duration(milliseconds: 300),
    this.curve = Curves.easeInOut,
    this.reverseDuration,
    this.reverseCurve,
    this.alignment,
    this.offsetFromRight = 60.0,
    this.rotateAngle = (pi / 24),
    this.isRotate = true,
    this.onWillPop,
  }) : super(key: key);

  static _SlideDrawerState? of(BuildContext context) =>
      context.findAncestorStateOfType<_SlideDrawerState>();

  @override
  _SlideDrawerState createState() => _SlideDrawerState();
}

class _SlideDrawerState extends State<SlideDrawer>
    with SingleTickerProviderStateMixin {
  bool _canBeDragged = false;
  double _minDragStartEdge = 60;
  double get _maxSlide =>
      MediaQuery.of(context).size.width - widget.offsetFromRight;
  double get _maxDragStartEdge => _maxSlide - 16;

  bool get _hasReverseCurve => widget.reverseCurve != null;
  bool get _hasReverseDuration => widget.reverseDuration != null;
  bool get _hasOnWillPop => widget.onWillPop != null;

  late CurvedAnimationController _animation;

  bool get isOpened => _animation.progress >= 1.0;
  bool get isClosed => _animation.value <= 0.0;

  @override
  void initState() {
    super.initState();
    _initAnimation();
  }

  @override
  void dispose() {
    _animation.dispose();
    super.dispose();
  }

  _initAnimation() {
    _animation = CurvedAnimationController(
      vsync: this,
      duration: widget.duration,
      curve: widget.curve,
      reverseCurve: _hasReverseCurve ? widget.reverseCurve : widget.curve,
      reverseDuration:
          _hasReverseDuration ? widget.reverseDuration : widget.duration,
      debugLabel: 'SlideDrawer',
    );

    _animation.addListener(() => setState(() {}));
  }

  reset() {
    _animation.dispose();
    _initAnimation();
  }

  open() => _animation.start();
  close() => _animation.reverse();
  toggle() => isOpened ? close() : open();

  _onDragStart(DragStartDetails details) {
    bool isDragOpenFromLeft =
        _animation.isDismissed && details.globalPosition.dx < _minDragStartEdge;

    bool isDragCloseFromRight =
        _animation.isCompleted && details.globalPosition.dx > _maxDragStartEdge;

    _canBeDragged = isDragOpenFromLeft || isDragCloseFromRight;
  }

  _onDragUpdate(DragUpdateDetails details) {
    if (_canBeDragged && details.primaryDelta != null) {
      double delta = details.primaryDelta! / _maxSlide;
      _animation.progress += delta;
    }
  }

  _onDragEnd(DragEndDetails details) {
    double _kMinFlingVelocity = 365.0;

    if (_animation.isDismissed || _animation.isCompleted) {
      return;
    }

    if (details.velocity.pixelsPerSecond.dx.abs() >= _kMinFlingVelocity) {
      double visualVelocity = details.velocity.pixelsPerSecond.dx /
          MediaQuery.of(context).size.width;

      _animation.fling(velocity: visualVelocity);
    } else if (_animation.value < 0.5) {
      close();
    } else {
      open();
    }
  }

  Widget get _drawer => SlideDrawerContainer(
        alignment: widget.alignment,
        brightness: widget.brightness,
        backgroundColor: widget.backgroundColor,
        backgroundGradient: widget.backgroundGradient,
        head: widget.headDrawer,
        content: widget.contentDrawer,
        drawer: widget.drawer,
        items: widget.items,
        paddingRight: widget.offsetFromRight,
      );

  Matrix4 get _transform => Matrix4.identity()
    ..translate(_maxSlide * _animation.value)
    ..scale(1.0 - (0.25 * _animation.value))
    ..setEntry(3, 2, 0.001)
    ..rotateY(_animation.value * widget.rotateAngle);

  Matrix4 get _transformNoRotate => Matrix4.identity()
    ..translate(_maxSlide * _animation.value)
    ..scale(1.0 - (0.25 * _animation.value));

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        if (isOpened) {
          close();
          return Future.value(false);
        } else if (_hasOnWillPop) {
          return widget.onWillPop?.call();
        } else {
          return Future.value(true);
        }
      },
      child: GestureDetector(
        onHorizontalDragStart: _onDragStart,
        onHorizontalDragUpdate: _onDragUpdate,
        onHorizontalDragEnd: _onDragEnd,
        child: Stack(
          children: [
            _drawer,
            Transform(
              alignment: Alignment.centerLeft,
              transform: widget.isRotate ? _transform : _transformNoRotate,
              child: Material(
                child: Stack(
                  children: [
                    widget.child,
                    if (isOpened)
                      GestureDetector(
                        child: Container(
                          width: MediaQuery.of(context).size.width,
                          height: MediaQuery.of(context).size.height,
                          color: Colors.white.withOpacity(0.01),
                        ),
                        onTap: close,
                      ),
                  ],
                ),
                elevation: 6,
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/container.dart
import 'package:flutter/material.dart';
import 'package:slide_drawer/src/alignment.dart';
import 'package:slide_drawer/src/item.dart';

import 'drawer.dart';

class SlideDrawerContainer extends StatelessWidget {
  final Widget? drawer;
  final Widget? head;
  final Widget? content;
  final List<MenuItem> items;
  final double paddingRight;

  /// The gradient to use for the background.
  ///
  /// If this property is null, then [background] is used.
  final Gradient? backgroundGradient;

  /// The color to use for the background. Typically this should be set
  /// along with [brightness].
  ///
  /// If this property is null, then [ThemeData.primaryColor] is used.
  final Color? backgroundColor;

  /// The brightness of the app bar's material. Typically this is set along
  /// with [backgroundColor], [backgroundGradient].
  ///
  /// If this property is null, then [ThemeData.primaryColorBrightness] is used.
  final Brightness? brightness;

  /// Vertical alignment of content inside [SlideDrawerContainer]
  /// it can [start] from the top, or [center]
  final SlideDrawerAlignment? alignment;

  bool get _hasItems => items.isNotEmpty;
  bool get _hasDrawer => drawer != null;
  bool get _hasHead => head != null;
  bool get _hasContent => content != null;
  bool get _hasGradient => backgroundGradient != null;

  SlideDrawerContainer({
    Key? key,
    required this.drawer,
    this.head,
    this.content,
    this.items = const [],
    this.brightness,
    this.backgroundColor,
    this.backgroundGradient,
    this.alignment,
    this.paddingRight = 0,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    ThemeData theme = Theme.of(context);
    SlideDrawerAlignment? _alignment = alignment;

    if (_alignment == null) {
      _alignment =
          _hasHead ? SlideDrawerAlignment.start : SlideDrawerAlignment.center;
    }

    bool _isAlignTop = _alignment == SlideDrawerAlignment.start;

    return Material(
      child: _hasDrawer
          ? drawer
          : Container(
              decoration: _hasGradient
                  ? BoxDecoration(gradient: backgroundGradient)
                  : BoxDecoration(color: backgroundColor ?? theme.primaryColor),
              child: SafeArea(
                child: Theme(
                  data: ThemeData(
                      brightness: brightness ?? theme.primaryColorBrightness),
                  child: Column(
                    mainAxisAlignment: _isAlignTop
                        ? MainAxisAlignment.start
                        : MainAxisAlignment.center,
                    crossAxisAlignment: CrossAxisAlignment.stretch,
                    children: [
                      if (_hasHead) head!,
                      if (_hasContent)
                        Container(
                          margin: EdgeInsets.only(right: paddingRight),
                          child: content,
                        ),
                      if (!_hasContent && _hasItems)
                        for (MenuItem item in items)
                          Container(
                            margin: EdgeInsets.only(right: paddingRight),
                            child: MenuItemWidget(item: item),
                          ),
                    ],
                  ),
                ),
              ),
            ),
    );
  }
}

class MenuItemWidget extends StatelessWidget {
  final MenuItem item;

  MenuItemWidget({Key? key, required this.item}) : super(key: key);

  Widget? get _leading {
    if (item.hasLeading) return item.leading!;
    if (item.hasIcon) return Icon(item.icon);
    return null;
  }

  @override
  Widget build(BuildContext context) {
    return ListTile(
      leading: _leading,
      contentPadding: EdgeInsets.only(left: _leading == null ? 24 : 16),
      title: Text(item.title),
      onTap: () {
        if (item.isCloseDrawerWhenTapped) {
          SlideDrawer.of(context)?.close();
        }

        item.onTap?.call();
      },
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:slide_drawer/slide_drawer.dart';

void main() {
  runApp(App());
}

class App extends StatefulWidget {
  @override
  _AppState createState() => _AppState();
}

enum SlideDrawerExampleType {
  BASIC,
  BACKGROUND,
  GRADIENT,
  HEAD_DRAWER,
  CONTENT_DRAWER,
  HEAD_CONTENT_DRAWER,
  FULL_DRAWER,
}

class _AppState extends State<App> {
  Key _appKey = UniqueKey();
  SlideDrawerExampleType type = SlideDrawerExampleType.BASIC;

  bool get _isBasic => type == SlideDrawerExampleType.BASIC;
  bool get _isBackground => type == SlideDrawerExampleType.BACKGROUND;
  bool get _isGradient => type == SlideDrawerExampleType.GRADIENT;
  bool get _isHead => type == SlideDrawerExampleType.HEAD_DRAWER;
  bool get _isContent => type == SlideDrawerExampleType.CONTENT_DRAWER;
  bool get _isHeadContent => type == SlideDrawerExampleType.HEAD_CONTENT_DRAWER;
  bool get _isFullDrawer => type == SlideDrawerExampleType.FULL_DRAWER;

  List<MenuItem> get _items => [
        MenuItem('Basic',
            onTap: () => _changeType(SlideDrawerExampleType.BASIC)),
        MenuItem('Custom Background',
            onTap: () => _changeType(SlideDrawerExampleType.BACKGROUND)),
        MenuItem('Custom Gradient',
            onTap: () => _changeType(SlideDrawerExampleType.GRADIENT)),
        MenuItem('Custom Header',
            onTap: () => _changeType(SlideDrawerExampleType.HEAD_DRAWER)),
        MenuItem('Custom Content',
            onTap: () => _changeType(SlideDrawerExampleType.CONTENT_DRAWER)),
        MenuItem('Header and Content',
            onTap: () =>
                _changeType(SlideDrawerExampleType.HEAD_CONTENT_DRAWER)),
        MenuItem('Full Drawer',
            onTap: () => _changeType(SlideDrawerExampleType.FULL_DRAWER)),
      ];

  List<MenuItem> get _itemsIcon => [
        MenuItem('Basic',
            icon: Icons.rss_feed,
            onTap: () => _changeType(SlideDrawerExampleType.BASIC)),
        MenuItem('Custom Background',
            icon: Icons.favorite_border,
            onTap: () => _changeType(SlideDrawerExampleType.BACKGROUND)),
        MenuItem('Custom Gradient',
            icon: Icons.mail_outline,
            onTap: () => _changeType(SlideDrawerExampleType.GRADIENT)),
        MenuItem('Custom Header',
            icon: Icons.map,
            onTap: () => _changeType(SlideDrawerExampleType.HEAD_DRAWER)),
        MenuItem('Custom Content',
            icon: Icons.person_outline,
            onTap: () => _changeType(SlideDrawerExampleType.CONTENT_DRAWER)),
        MenuItem('Header and Content',
            icon: Icons.alarm,
            onTap: () =>
                _changeType(SlideDrawerExampleType.HEAD_CONTENT_DRAWER)),
        MenuItem('Full Drawer',
            icon: Icons.settings,
            onTap: () => _changeType(SlideDrawerExampleType.FULL_DRAWER)),
      ];

  _changeType(type) {
    if (this.type != type)
      setState(() {
        this.type = type;
        _appKey = UniqueKey();
      });
  }

  Widget get _home {
    if (_isBasic) return _basic;
    if (_isBackground) return _background;
    if (_isGradient) return _gradient;
    if (_isHead) return _head;
    if (_isContent) return _content;
    if (_isHeadContent) return _headContent;
    if (_isFullDrawer) return _fullDrawer;
    return _basic;
  }

  Widget get _basic => SlideDrawer(
        child: HomePage(),
        items: _items,
      );

  Widget get _background => SlideDrawer(
        child: HomePage(),
        brightness: Brightness.dark,
        backgroundColor: Colors.blue[900],
        items: _items,
      );

  Widget get _gradient => SlideDrawer(
        child: HomePage(),
        brightness: Brightness.dark,
        backgroundGradient: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          stops: [0.0, 1.0],
          colors: [
            Color(0xFF000046),
            Color(0xFF1CB5E0),
          ],
        ),
        curve: Curves.easeInOut,
        items: _itemsIcon,
      );

  Widget get _head => SlideDrawer(
        child: HomePage(),
        brightness: Brightness.dark,
        backgroundColor: Colors.blue[800],
        headDrawer: Image.asset('boys.png'),
        alignment: SlideDrawerAlignment.start,
        items: _itemsIcon,
      );

  Widget get _content => SlideDrawer(
        child: HomePage(),
        brightness: Brightness.dark,
        contentDrawer: Container(
          padding: EdgeInsets.symmetric(horizontal: 5),
          child: Column(
            children: [
              for (var item in _itemsIcon)
                ListTile(
                  title: Text(item.title),
                  leading: Icon(item.icon),
                  onTap: item.onTap,
                ),
            ],
          ),
        ),
        backgroundGradient: LinearGradient(
          begin: Alignment.topCenter,
          end: Alignment.bottomCenter,
          stops: [0.0, 1.0],
          colors: [
            Colors.blue[800]!,
            Color(0xFF1CB5E0),
          ],
        ),
      );

  Widget get _headContent => SlideDrawer(
        child: HomePage(),
        brightness: Brightness.light,
        backgroundColor: Color(0xffededed),
        alignment: SlideDrawerAlignment.start,
        headDrawer: Image.asset('boys.png'),
        contentDrawer: Container(
          padding: EdgeInsets.only(left: 10, top: 18),
          child: Column(
            children: [
              for (var item in _itemsIcon)
                ListTile(
                  title: Text(item.title),
                  leading: Icon(item.icon),
                  onTap: item.onTap,
                ),
            ],
          ),
        ),
      );

  Widget get _fullDrawer => SlideDrawer(
        child: HomePage(),
        isRotate: false,
        offsetFromRight: 100,
        drawer: Container(
          color: Colors.teal,
          padding: EdgeInsets.symmetric(vertical: 36, horizontal: 15),
          child: Theme(
            data: ThemeData(brightness: Brightness.dark),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.end,
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                for (var item in _items)
                  ListTile(
                    title: Text(item.title),
                    onTap: item.onTap,
                  ),
              ],
            ),
          ),
        ),
      );

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      key: _appKey,
      title: 'Slide Drawer Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: _home,
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text('Slide Drawer Demo'),
        leading: IconButton(
          icon: Icon(Icons.menu),
          // call toggle from SlideDrawer to alternate between open and close
          // when pressed menu button
          onPressed: () => SlideDrawer.of(context)?.toggle(),
        ),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
